risc0/zkvm/platform/io.h

```diff
@@ -16,54 +16,114 @@
 
 namespace risc0 {
 
+// These must match the values in zkvm/sdk/rust/guest/src/gpio.rs
 constexpr size_t kGPIO_SHA = 0x001C0000;
-constexpr size_t kGPIO_Write = 0x001C0004;
-constexpr size_t kGPIO_Commit = 0x001C0008;
-constexpr size_t kGPIO_Fault = 0x001C000C;
-constexpr size_t kGPIO_Log = 0x001C0010;
-constexpr size_t kGPIO_GetKey = 0x001C0014;
+constexpr size_t kGPIO_Commit = 0x001C0004;
+constexpr size_t kGPIO_Fault = 0x001C0008;
+constexpr size_t kGPIO_Log = 0x001C000C;
+constexpr size_t kGPIO_GetKey = 0x001C010;
 
-struct ShaDescriptor;
+constexpr size_t kGPIO_SendRecvChannel = 0X001C0014;
+constexpr size_t kGPIO_SendRecvSize = 0x001C0018;
+constexpr size_t kGPIO_SendRecvAddr = 0x001C001C;
 
-struct IoDescriptor {
-  uint32_t size;
-  uint32_t addr;
-};
+// Standard ZKVM channels; must match zkvm/sdk/rust/guest/src/gpio.rs.
 
-struct FaultDescriptor {
-  uint32_t addr;
-};
+// Request the initial input to the guest.
+constexpr uint32_t kSendRecvChannel_InitialInput = 0;
+// Write bytes to standard output
+constexpr uint32_t kSendRecvChannel_Stdout = 1;
 
-struct LogDescriptor {
-  uint32_t addr;
-};
+// To invoke accelerated SHA, the guest writes ShaDescriptor structs
+// in sequence to the "SHA" memory region.  Once the ShaDescriptor has
+// been written, it writes a pointer to the written ShaDescriptor
+// structure to kGPIO_SHA.
+//
+// The guest writes a single "0" word (in what would be the
+// "typeAndCount" field of the next ShaDescriptor structure) after all
+// the ShaDescriptor structures in the SHA region before exiting.
+//
+// Upon exit, the guest's ZKP Sha implementation looks through the
+// ShaDescriptor space and verifies all SHAs that were computed during
+// the run.  Due to this, any memory
+struct ShaDescriptor {
+  // Number of 64-byte chunks of SHA data to process
+  uint32_t typeAndCount;
 
-struct GetKeyDescriptor {
-  uint32_t name;
-  uint32_t addr;
-  uint32_t mode;
+  // TODO(nils): "idx" seems to be always zero; is it used for anything?
```

Comment:
This was intended for the merkle index, where tha SHA verifiers would have a build in 'merkle path verifiers' since it saves quite a bit of work and for recursion merkle verifications at literally 90% of the total compute, never finished this portion of the circuit.